helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -401,6 +402,8 @@ void checkConnected(long timeout) {
       throw new HelixException(
           "HelixManager is not connected within retry timeout for cluster " + _clusterName);
     }
+
+    _sessionId = ZKUtil.toHexSessionId(_zkclient.getSessionId());
```

Comment:
This may cause some issues. If the session is updated randomly, a new session may be populated before new session handling is done.

Reply:
I was waiting for you to point it out and discuss :) I had the same thought when I considered adding this assignment. _sessionId is only assigned in waitUntilConnected(). Each time handleNewSession() also calls waitUntilConnected(). handleNewSession(expectedSession) handles expected session. If the sessionId changes before handleNewSession() is done, I expect handleNewSession() should be able to handle the expired session because it is handling an expected session? Maybe I miss some corner cases? Let's carefully think about this here.